### PR TITLE
Use Scene.onBeforeRender() and onAfterRender() in WebGLRenderer for tiled forward lighting and VR post proc

### PR DESCRIPTION
**Pull request by wizgrav**
Use Scene.onBeforeRender on WebGLRenderer to provide a way to get the vr cameras after they are updated for the current frame toimplement forward+ shading and expose the vr cameras for post processing.

### CODE REVIEW

**Comment by edankwan** [comment]
@wizgrav can you update and fix the conflict? @mrdoob it should be a simple update. Please offer us some love for VR post effect 🤗

**Comment by wizgrav** [comment]
@edankwan done, it was a silly conflict with a comment(?) @mrdoob the changes are really minimal and still relevant after all this time. It won't just help with post proc. The updated cameras are also needed to construct the textures needed for forward+ shading, after the cameras update and just before the rendering of other objects is performed

**Comment by takahirox** [comment]
I don't have strong preference but `renderer.vr.autoSubmitFrame` sounds a bit better than `renderer.autoSubmitFrame` because users can easily speculate `.autoSubmitFrame` is for VR?

@mrdoob 

If you have some concerns for `renderer.onBeforeRender()` while `.autoSubmitFrame` seems ok to you, can we get `.autoSubmitFrame` in first because we need it for PostProcessing on VR?

**Comment by mrdoob** [comment]
Hmm... This feels pretty hacky. Can anyone share a snippet of how the code would be used?

**Comment by wizgrav** [comment]
```js
renderer.onBeforeRender = function (renderer, scene, camera, renderTarget) {
  // Use camera to manually cull lights, generate the per screen tile light lists and store them in texture(s)
};
```
These textures will then be used by custom materials to implement Forward+ shading as described here [Forward-Plus-Renderer](https://github.com/bcrusco/Forward-Plus-Renderer)

Its like doing many lights in deffered, but generally faster since we don't always have mrt support.

The reason onBeforeRender is needed there is because we need to tap on the camera after it updates. In WebGLRenderer.render the camera can also be swapped with the vr ArrayCamera, so by tapping where it does, onBeforeRender will be called with the correct camera that will be used for the scene rendering

**Comment by takahirox** [comment]
Here is the very basic PostProcessing + VR example code.

    renderer.vr.enabled = true;
    renderer.vr.autoSubmitFrame = false;

    scene.add( mesh );

    mesh2.material.uniforms.tDiffuse.value = renderTarget.texture;
    scene2.add( mesh2 );

    function animate() {

        requestAnimationFrame( animate );

        // main rendering to renderTarget
        renderer.render( scene, camera, renderTarget );

        // post-processing
        // renderer.vr.enabled = false;  // some post-processing might require non-VR mode
        renderer.render( scene2, camera2 );
        // renderer.vr.enabled = true;

        renderer.submitFrame();

    }

**Comment by wizgrav** [comment]
@takahirox  @mrdoob Also some post proc effects need the camera(s) used to render the original. Godrays would be an example of that which would need to calculate the center for the radial blur. onBeforeRender would come in handy here as well

**Comment by mrdoob** [comment]
Hmm... How about something like `renderer.vr.onBeforeSubmitFrame = function() {}`?

**Comment by wizgrav** [comment]
@mrdoob .onBeforeSubmitFrame() would be right before the submitFrame() call right? In this case the scene will be already rendered by the time it is called so the tiled forward technique cannot be implemented. The tiled light textures need to be created just after the cameras have updated but before the scene objects are rendered because their shaders need the up to date textures to perform lighting.

Another proposal would be onBeforeRender() where it is and an onAfterRender() just before the frame is submitted. This way post proc can be performed in the latter and it will work for both normal or vr mode.  This adds another method but takes out the autoSubmitFrame flag and is symmetric.

**Comment by mrdoob** [comment]
> Another proposal would be onBeforeRender() where it is and an onAfterRender() just before the frame is submitted. This way post proc can be performed in the latter and it will work for both normal or vr mode.

Hmm, do you mind pointing me to the exact lines in the code where we would be calling these functions?

**Comment by mrdoob** [comment]
Oh wait, you already modified the PR.

File: src/renderers/WebGLRenderer.js [hunk 1]
```diff
@@ -1099,6 +1103,16 @@ function WebGLRenderer( parameters ) {
 
 		}
 
+		var tempHandler = _this.onBeforeRender;
+
+		if ( tempHandler ) {
+
+			_this._onBeforeRender = null;
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: I would put this on your app code instead. We do similar things when using `object.onBeforeRender` already:

https://github.com/mrdoob/three.js/blob/dev/examples/js/objects/Reflector.js#L158-L162

**Comment by wizgrav** [comment]
@mrdoob I reverted the code, I guess a warning in the docs should suffice for avoiding the infinite recursion and I didn't like the tempHandler var in the renderer. In the weekend, I'll get on to making an example for tiled forward lighting and post proc

**Comment by edankwan** [comment]
I personally prefer `.autoSubmitFrame` approach because of the infinite recursion issue @wizgrav mentioned. In order to bypass this issue, we will need to do some hacky stuff which I don't think it is a good solution.

**Comment by wizgrav** [comment]
Happy new year every one, wishes for a prosperous and creative 2018 :) 

I finally found some time to write an example for tiled forward lighting. It's probably the most basic implementation, using tile texels as bitmasks to accomodate 32 lights on screen. It can be easily extended to more, though it would not be best approach for more than a couple hundred. There are other approaches using one more intermediate texture that can accommodate thousands of lights.

I added back the autoSubmitFrame flag as @edankwan suggested. I used the onAfterRender() approach for the demo and the mentioned hack for the infinite recursion is visible and not very pleasing imho. Maybe we could keep both approaches as the code footprint is really tiny. After considering @takahirox suggestion, I decided to leave the flag on the renderer and not the VRManager as the renderer is the one actually submitting the frames so it makes more sense to me that he should carry the toggle.

**Comment by wizgrav** [comment]
And here's the example live https://wizgrav.github.io/three.js/examples/webgl_tiled_forward.html I need to cleanup the build files for this PR. Your thoughts @mrdoob ?

File: src/renderers/WebGLRenderer.js [hunk 2]
```diff
@@ -1099,6 +1105,8 @@ function WebGLRenderer( parameters ) {
 
 		}
 
+		_this.onBeforeRender( _this, scene, camera, renderTarget );
```
Comment (pailhead) [inline src/renderers/WebGLRenderer.js]: could this possible be `onAfterAutoUpdate` ?

**Comment by pailhead** [comment]
It seems to me that the only thing special about the `onBeforeRender` call is that it may happen after the scene autoUpdate. 

Making this call before you actually call `myRenderer.render()` seems to guarantee only these two things:
1) the scenegraph objects will have their matrices up to date, if they were set to auto update
2) the render camera will have a correct matrix if it does not belong to the scenegraph (no parent)

What I was suggesting before was to have an object based `onBeforeUpdate` and `onAfterUpdate` that would be called before and after the matrices are computed. It feels like it would give more meaning to:

```
if ( scene.autoUpdate === true ) scene.updateMatrixWorld();
```

I'm not sure if i explained it well, since i didn't explain it well the first time i was suggesting this. 

With how it's currently written/named it seems that this would make more sense:

```

function actualBeforeRenderCall(){...}

myScene.autoUpdate = false

function myRender(){
   myScene.updateMatrixWorld(true)
   actualBeforeRenderCall()
   myRenderer.render(myScene, myCamera)
}
```

**Comment by wizgrav** [comment]
@pailhead in the case of vr, the provided camera is replaced by an, internally updated, ArrayCamera, This is the only reason onBeforeRender is needed, It provides a convenient structure for both cases, vr or non. Same goes for onAfterRender. Some post proc effects also need the camera(s)

**Comment by wizgrav** [comment]
@mrdoob changed the location of the callbacks to be on the Scene object which supported them already from Object3D. I decided to leave .autoSubmitFrame out as, with this arrangement, the renderer can be reused safely inside onAfterRender @edankwan.

File: examples/webgl_tiled_forward.html [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+
+		var State = {
+			rows:0,
+			cols:0,
+			width: 0,
+			height: 0,
+			tileData : { type: "v4", value: null },
+			tileTexture: { type: "t", value: null },
+			lightTexture: {
+				type: "t",
```
Comment (WestLangley) [inline examples/webgl_tiled_forward.html]: `type` specification should not be needed.

File: examples/webgl_tiled_forward.html [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+					transparent: tIndex === index ? true : false,
+
+				});
+				mtl.uniforms.opacity.value = tIndex === index ? 0.9 : 1;
+				mtl.uniforms.tileData = State.tileData;
+				mtl.uniforms.tileTexture = State.tileTexture;
+				mtl.uniforms.lightTexture = State.lightTexture;
+				for( var u in conf.uniforms ) {
+					var vu = conf.uniforms[u];
+					if(mtl.uniforms[u].value.set) {
```
Comment (WestLangley) [inline examples/webgl_tiled_forward.html]: maybe
```javascript
if ( mtl.uniforms[ u ].value && mtl.uniforms[ u ].value.set ) {
```

File: examples/webgl_tiled_forward.html [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+			State.width = width;
+			State.height = height;
+			State.cols = Math.ceil(width / 32);
+			State.rows = Math.ceil(height / 32);
+			State.tileData.value = [ width, height, 0.5 / Math.ceil( width / 32), 0.5 / Math.ceil( height / 32) ];
+			State.tileTexture.value = new THREE.DataTexture( new Uint8Array(State.cols * State.rows * 4), State.cols, State.rows);
+		}
+
+		// Generate the light bitmasks and store them in the tile texture
+		function tileLights(renderer, scene, camera) {
```
Comment (WestLangley) [inline examples/webgl_tiled_forward.html]: Can you clean up your whitespace formatting a bit? https://zz85.github.io/mrdoobapproves/

File: examples/webgl_tiled_forward.html [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+					l.add(new THREE.Mesh(
+						sphereGeom, 
+						new THREE.MeshBasicMaterial( { color: new THREE.Color("hsl(" + chroma + ", " + sat + "%, 50%)") } )
+					));
+					l.add(new THREE.Mesh(
+						sphereGeom, 
+						new THREE.MeshBasicMaterial( { 
+							color: new THREE.Color("hsl(" + chroma + ", " + sat + "%, 50%)"), 
+							transparent: true,
+							opacity: 0.033
```
Comment (WestLangley) [inline examples/webgl_tiled_forward.html]: is that even visible?

File: examples/webgl_tiled_forward.html [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+			"  		getPointDirectLightIrradiance( pointlight, geometry, directLight );",
+			"		RE_Direct( directLight, geometry, material, reflectedLight );",
+			"	}",
+			"}",
+			"#endif"
+		].join("\n");
+
+		var lights = [], objects = [];
+
+		var State = {
```
Comment (WestLangley) [inline examples/webgl_tiled_forward.html]: lower case

**Comment by WestLangley** [comment]
>changed the location of the hooks to be on the Scene object which supported them already

Much better.  :-)

I have not been able to study the tiled-forward code yet, but this is an interesting demo of a shader hack.

It is also an interesting implementation of post processing in lieu of `EffectComposer`.

File: examples/webgl_tiled_forward.html (hunk 6 above)
Comment (wizgrav) [reply examples/webgl_tiled_forward.html]: Barely, it makes it look like a cheap glow around the light spheres

**Comment by mrdoob** [comment]
Love how simple the PR turned out to be! 😍

**Comment by mrdoob** [comment]
The example definitely needs some clean up, but I'll merge this for now.

**Comment by mrdoob** [comment]
Thanks!

File: examples/webgl_tiled_forward.html (hunk 6 above)
Comment (WestLangley) [reply examples/webgl_tiled_forward.html]: It's nice. On my monitor I had to increase the opacity, though.

**Comment by takahirox** [comment]
Finally we can start to work on VR + Post-Proc? 😄

**Comment by wizgrav** [comment]
@WestLangley yeah, I basically did them but mrdoob merged first. I was also thinking to vr enable the example for completeness

**Comment by mrdoob** [comment]
@wizgrav feel free create another PR with these changes 😊

**Comment by DusanBosnjakKodiak** [comment]
@looeee 

I believe this stuff is related to your post on the forum.

**Pull request merged**